### PR TITLE
   [sub-plugin] Modify synchronization mechanism between push_data and getSample

### PR DESCRIPTION
**Pull request by songgot**
Modify synchronization mechanism between push_data and getSample
    - Remove some member variable
    - Add member function to check queue

**Self evaluation:**
1. Build test:   [X]Passed [ ]Failed [ ]Skipped
2. Run test:     [X]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

File: nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc [hunk 1]
```diff
@@ -182,7 +182,9 @@ static int nntrainer_model_push_data(const GstTensorTrainerFramework *fw,
   }
 
   data->push_count++;
+  data->count_lock.lock();
   data->queue_count++;
+  data->count_lock.unlock();
```
Comment (myungjoo) [inline nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc:187]: It is strange to protect INC(queue_count) while you don't protect READ(queue_count) or DEC(queue_count) with the lock. 

Please check the critical sections and protected data. Locking queue_count++ only doesn't look meaningful. (and consider using "atomic")

File: nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc [hunk 2]
```diff
@@ -225,6 +227,7 @@ void NNTrainer::InputTensorsInfo::getSample(float **input, float **label,
     is_data_wait_locked = TRUE;
     data_wait.wait(lock);
     ml_logd("unlocked, get data");
+    is_data_wait_locked = FALSE;
```
Comment (myungjoo) [inline nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc:230]: is_data_wait_locked is not protected. are you sure you can use it like this?

**Review by myungjoo — CHANGES_REQUESTED** [review]
Reconsider the synchronization mechanisms.
The usage of count_lock is very strange.
The usage of is_data_full_locked is also strange. (in general, you cannot/shouldn't check the lock status with an unprotected scalar variable.)

**Comment by songgot** [comment]
Thank you for your review, I will change PR :)

**Comment by songgot** [comment]
PR is changed, Thank you.

File: nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc (hunk 1 above)
Comment (songgot) [reply nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc]: PR is changed, Thank you.

File: nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc (hunk 2 above)
Comment (songgot) [reply nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc]: PR is changed, Thank you.

File: nnstreamer/tensor_trainer/tensor_trainer_nntrainer.hh [hunk 3]
```diff
@@ -141,17 +142,13 @@ public:
...
   unsigned int queue_rear;
-  unsigned int queue_count; /**< The number of data in queue */
-  unsigned int push_count;  /**< The number of samples pushed to queue by
-                               NNStreamer(tensor_trainer) */
-  unsigned int pop_count;   /**< The number of pop from the queue for pushing
-                               samples to nntrainer */
+  unsigned int push_count; /**< The number of samples pushed to queue by
+                              NNStreamer(tensor_trainer) */
+  unsigned int pop_count;  /**< The number of pop from the queue for pushing
+                              samples to nntrainer */
```
Comment (myungjoo) [inline nnstreamer/tensor_trainer/tensor_trainer_nntrainer.hh:151]: Leave "API" methods only for public. Make all the others private or protected. For C API function accessing private info, add friend. Don't leave private properties public, future developers will abuse them.

Comment (songgot) [reply nnstreamer/tensor_trainer/tensor_trainer_nntrainer.hh:151]: Thank you for your detailed review, I will update PR ASAP.
Thank you.

File: nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc [hunk 4]
```diff
@@ -182,29 +209,17 @@ static int nntrainer_model_push_data(const GstTensorTrainerFramework *fw,
...
+  int queue_count = (data->queue_rear - data->queue_front + data->queue_size) %
+                    data->queue_size;
 
+  ml_logd("front:%d, rear:%d, filled:%d", data->queue_front, data->queue_rear,
+          queue_count);
   ml_logd("(pop/push: %d/%d)", data->pop_count, data->push_count);
   ml_logd("T-pushed: %d/%d, V-pushed:%d/%d\n",
           nntrainer->train_data->push_count, nntrainer->num_training_samples,
           nntrainer->valid_data->push_count, nntrainer->num_validation_samples);
+  data->data_empty.notify_one();
```
Comment (myungjoo) [inline nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc:222]: Reference: https://en.cppreference.com/w/cpp/thread/condition_variable

```lock.unlock()``` should be done before notify_one

File: nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc [hunk 5]
```diff
@@ -260,24 +267,9 @@ void NNTrainer::InputTensorsInfo::getSample(float **input, float **label,
...
-
-  /* to avoid dead lock, check is_data_full_locked */
-  if (!is_data_full_locked && queue_count == 0) {
-    ml_logd("locked, need to wait for more data");
-    std::unique_lock<std::mutex> lock(data_wait_lock);
-    is_data_wait_locked = TRUE;
-    data_wait.wait(lock);
-    ml_logd("unlocked, get data");
-  }
+  data_full.notify_one();
```
Comment (myungjoo) [inline nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc:272]: ```lock.unlock()``` missing before notify_one

**Comment by myungjoo** [comment]
Design of the class InputTensorsInfo looks weird.

If it has getSample (pop), it should also have putSample (push). Having push function out of the class (nntrainer_model_push_data) doesn't look appropriate. It is also against general OO design principles.

**Comment by songgot** [comment]
> Design of the class InputTensorsInfo looks weird.
> 
> If it has getSample (pop), it should also have putSample (push). Having push function out of the class (nntrainer_model_push_data) doesn't look appropriate. It is also against general OO design principles, causing more potential bugs. Please try to put all the properties private and let them accessed via methods, unless the properties are trivial; locks cannot be public.
> 
> The C functions in GstTensorTrainerFramework should not be complex. They should be trivial wrappers calling class member methods.
Thank you. 
getSample is cb of nntrainer.
When nntrainer needs data, it calls cb.

File: nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc (hunk 5 above)
Comment (songgot) [reply nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc:272]: data_full's lock is in nntrainer_model_push_data(), if isQueueFull() is TRUE, locked
```
  std::unique_lock<std::mutex> lock(data->queue_lock);
  data->data_full.wait(lock, [&] { return !data->isQueueFull(); });
 ```
data_full.notify_one() is called ``` data->data_full.wait(lock, [&] { return !data->isQueueFull(); });``` is unlocked

File: nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc (hunk 4 above)
Comment (songgot) [reply nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc:222]: The structure is as follows. 
Does this mean that queue_front and queue_rear require a mutex?
```
 std::mutex queue_lock;
 std::condition_variable data_full;
 std::condition_variable data_empty;

getSample() { // use queue data

  std::unique_lock<std::mutex> lock(queue_lock);
  data_empty.wait(lock, [this] { return !isQueueEmpty(); }); // Lock when the queue is empty
  [...]
  queue_front = (queue_front + 1) % queue_size;
  data_full.notify_one() // after pop data
}

push_data() { // write queue data
{
  std::unique_lock<std::mutex> lock(data->queue_lock);
  data->data_full.wait(lock, [&] { return !data->isQueueFull(); }); // Lock when the queue is full
 [...]
  data->queue_rear = (data->queue_rear + 1) % data->queue_size;
  data->data_empty.notify_one(); // after push data
}
```

**Comment by songgot** [comment]
You are a really good person(@myungjoo).
Thank you very much for your detailed explanation offline.

**Review by myungjoo — APPROVED** [review]

**Review by DonghakPark — APPROVED** [review]
LGTM!

**Review by djeong20 — APPROVED** [review]
nice thread-safe implementation!
any reason why you chose to use CV instead of semaphore?

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Pull request merged**